### ISSUE TITLE
Discussion

### ISSUE THREAD

**Issue by 2426021684** [2426021684/nlp-compromise-typings#1]
@eshansingh I am awake now

**Comment by eshansingh** [comment]
The API reference is _useless_! I haven't made much progress. For almost 90% of the day I've been wading though `nlp_comprimise`'s horribly undocumented code. There's barely any comments at all! And I haven't been able to get any of my other work done either.

**Comment by eshansingh** [comment]
This is pretty much a futile effort until we understand the code better. The variable names are all over the place, they don't offer any self-documenting value. We need to try and figure this out.

**Comment by 2426021684** [comment]
I am just running the code in a repl to figure out the argument and return types. Sometimes I can infer it from the code.

**Comment by 2426021684** [comment]
https://www.npmjs.com/package/tsdoc could be an option

**Comment by eshansingh** [comment]
@2426021684 Good idea. Node REPL, you mean?

**Comment by 2426021684** [comment]
@eshansingh I am using a node repl called Mancy. https://github.com/princejwesley/Mancy

**Comment by eshansingh** [comment]
@2426021684 Whoa! Didn't know this existed. I prefer to work in command line, but that's awesome-looking!

**Comment by 2426021684** [comment]
@eshansingh why did you change `this` to `Sentence` for expand and contract? If you negate a question it should still be a question right?

**Comment by eshansingh** [comment]
@2426021684 I agree with that, but since that was in a `contractions` object, it referred to that contraction object. In addition, `tsc` complained with an error.

**Comment by 2426021684** [comment]
@eshansingh Ok

**Comment by eshansingh** [comment]
@2426021684 Another option is to have a temporary variable `_this` or something, but I'm not sure that would work because then the method would just return the initial state again (I think, haven't written a definitions file before even though I know TypeScript). Can you help/correct me on this?

**Comment by 2426021684** [comment]
@eshansingh maybe have contractions be a generic interface?

**Comment by 2426021684** [comment]
@eshansingh why should the typings for `Person.gender()` be?
I think the code returns only "Male", "Female", or null.

**Comment by eshansingh** [comment]
@2426021684 Having a whole separate interface for that _seems_ overkill, but that might be the only option. Know of any good references for definition files?

**Comment by eshansingh** [comment]
@2426021684 It would be a good enough thing to assume. Can we add @spencermountain here?

**Comment by eshansingh** [comment]
@2426021684 However, you can't use `"Male" | "Female" | null` for that, as again, `tsc` complains. I changed it to `"Male" | "Female" | any`. Not optimal, but again, I don't think either of us know the proper way. I have some other work to do, can you search for some help/docs?

**Comment by 2426021684** [comment]
@eshansingh Ok

**Comment by 2426021684** [comment]
@eshansingh `"Male" | "Female" | null` works fine for me. `"Male" | "Female" ` allows null too.

**Comment by eshansingh** [comment]
@2426021684 Weird. We'll use `"Male" | "Female"` as it is clearer as well.

**Comment by 2426021684** [comment]
Should we just use the existing tests for nlp-compromise? Or should we write our own.

**Comment by 2426021684** [comment]
Everything is typed that I know of. @spencermountain can you check everything especially POS and the tags. If everything is fine I will create a PR.

**Comment by 2426021684** [comment]
@spencermountain what does Result.match return?
